### ISSUE TITLE
Auto generated init-models file throws AssociationError

### ISSUE THREAD

**Issue by KjellVermeersch** [sequelize/sequelize-auto#449]
While generating a fresh init-models file, I have noticed that it always throws an AssociationError when running the freshly generated code.

Right now, the generated code is applying the through tabel to the association via the model itself. It looks like the init-model file contains a belongsTo and a hasMany, together with the belongsToMany associations. Which throws the error (I think). 
When removing the belongsTo and hasMany associations, the code runs fine (as far as I know). This issue occured both on plain JavaScript and TypeScript.

Is it possible to get this fixed, or do I as a user need to do something to stop this behaviour?

**The error:**
`
SequelizeAssociationError: You have used the alias PlanHasSpecifications in two separate associations. Aliased associations must have unique aliases.
`

**Example of associations in init-models file:**

```JS
PlanHasSpecifications.belongsTo(Plans, { foreignKey: "planName"});
PlanSpecifications.belongsToMany(Plans, { through: PlanHasSpecifications, foreignKey: "specificationName", otherKey: "planName" });
Plans.hasMany(PlanHasSpecifications, { foreignKey: "planName"});
PlanHasSpecifications.belongsTo(PlanSpecifications, { foreignKey: "specificationName"});
Plans.belongsToMany(PlanSpecifications, { through: PlanHasSpecifications, foreignKey: "planName", otherKey: "specificationName" });
PlanSpecifications.hasMany(PlanHasSpecifications, { foreignKey: "specificationName"});
```
**Example of (temporarily) fixed code:**
```JS
//PlanHasSpecifications.belongsTo(Plans, { foreignKey: "planName"});
PlanSpecifications.belongsToMany(Plans, { through: PlanHasSpecifications, foreignKey: "specificationName", otherKey: "planName" });
//Plans.hasMany(PlanHasSpecifications, { foreignKey: "planName"});
//PlanHasSpecifications.belongsTo(PlanSpecifications, { foreignKey: "specificationName"});
Plans.belongsToMany(PlanSpecifications, { through: PlanHasSpecifications, foreignKey: "planName", otherKey: "specificationName" });
//PlanSpecifications.hasMany(PlanHasSpecifications, { foreignKey: "specificationName"});
```

Thanks in advance.

**Comment by steveschmitt** [comment]
Yes, it looks like we could exclude the relations that are superseded by other `belongsToMany` relations.  But it's hard to tell when that is appropriate in general.  The alternative is to supply an alias for some of the relations.

Can you try fixing the code by adding a unique "as" property to the relations, such as:
```js
PlanHasSpecifications.belongsTo(Plans, { as: "hasPlan", foreignKey: "planName"});
PlanSpecifications.belongsToMany(Plans, { as: "specPlan", through: PlanHasSpecifications, foreignKey: "specificationName", otherKey: "planName" });
...
```

**Comment by KjellVermeersch** [comment]
According to [this page](https://gist.github.com/elliette/20ddc4e827efd9d62bc98752e7a62610#some-important-addendums), belongsToMany and hasMany should not be declaired together. I guess the same is true for belongsTo.

> 1. You should not use use belongsToMany and hasMany together, because they do fundamentally different things in regards to SQL. In our film-festival example, you could say a Festival.hasMany(Films) or that a Film.belongsToMany(Festival, {through: FestivalFilm} ) but you should never write both.

Best case should be to remove the unnecessary associations when working with a many-to-many relationship.
But as you asked, you can use different aliases to (temporarily) remove this issue. You could for example set the alias for these associations as following:
```JS
PlanHasSpecifications.belongsTo(Plans, { as: 'planHasSpecifications1', foreignKey: "planName"});
PlanSpecifications.belongsToMany(Plans, { through: PlanHasSpecifications, foreignKey: "specificationName", otherKey: "planName" });
Plans.hasMany(PlanHasSpecifications, { as: 'planHasSpecifications', foreignKey: "planName"});
PlanHasSpecifications.belongsTo(PlanSpecifications, { as: 'planHasSpecifications2', foreignKey: "specificationName"});
Plans.belongsToMany(PlanSpecifications, { through: PlanHasSpecifications, foreignKey: "planName", otherKey: "specificationName" });
PlanSpecifications.hasMany(PlanHasSpecifications, { as: 'planHasSpecifications', foreignKey: "specificationName"});
```

In this example, the alias for every belongsTo is set as the name from the table (according to the configuration), followed by an indexed number, in this case 1 or 2.
For the hasMany associations, you could just use the same name (as far as I know), but this name should be different than the model name and is independent from the belongsTo aliases.

This approach should only be applied to the hasMany's and belongsTo's that are added for many-to-many relationships to not confuse the users, I guess.

There are propably different ways to fix or get around this issue, but I think that this is a simple way to (temporarily) fix it. I have not yet been able to test this fully, but it looks like it's working.

I hope you can do something with this information, or get your hands on a better solution. 
Keep me in touch!

**FYI:**
As I noticed, the error does occures when there are two belongsTo associations with the same name. HasMany can have the same aliases, but need to be different from the belongsToMany assocications.

**Comment by KjellVermeersch** [comment]
After taking a quick look at the code in the repo, isn't it possible to stop the belongsTo and hasOne/hasMany associations from being added to the output if there is a many-to-many relationship?

In the file [auto-writer.ts](https://github.com/sequelize/sequelize-auto/blob/master/src/auto-writer.ts) already is a check for it. If you would add some extra checking, shouldn't it be possible to fix this issue?

The code could look something like this (a bit "ugly" because of the copied str append):
```JS
/** Create the belongsTo/hasMany/hasOne association strings */
private createAssociations() {
  let str = "";
  const fkTables = _.keys(this.foreignKeys).sort();
  fkTables.forEach(t => {
    const [schemaName, tableName] = qNameSplit(t);
    const modelName = recase(this.options.caseModel, tableName);
    const fkFields = this.foreignKeys[t];
    const fkFieldNames = _.keys(fkFields);
    fkFieldNames.forEach(fkFieldName => {
      const spec = fkFields[fkFieldName];
      if (spec.isForeignKey) {
        const targetModel = recase(this.options.caseModel, spec.foreignSources.target_table as string);
        const sourceProp = recase(this.options.caseProp, fkFieldName);

        // use "hasOne" cardinality if this FK is also a single-column Primary or Unique key; else "hasMany"
        const isOne = ((spec.isPrimaryKey && !_.some(fkFields, f => f.isPrimaryKey && f.source_column !== fkFieldName) ||
          (spec.isUnique && !_.some(fkFields, f => f.isUnique === spec.isUnique && f.source_column !== fkFieldName))));
        const hasRel = isOne ? "hasOne" : "hasMany";

        if (spec.isPrimaryKey) {
          // if FK is also part of the PK, see if there is a "many-to-many" junction
          const otherKey = _.find(fkFields, k => k.isForeignKey && k.isPrimaryKey && k.source_column !== fkFieldName);
          if (otherKey) {
            const otherModel = recase(this.options.caseModel, otherKey.foreignSources.target_table as string);
            const otherProp = recase(this.options.caseProp, otherKey.source_column);
            str += `  ${otherModel}.belongsToMany(${targetModel}, { through: ${modelName}, foreignKey: "${otherProp}", otherKey: "${sourceProp}" });\n`;
          } else {
            str += `  ${modelName}.belongsTo(${targetModel}, { foreignKey: "${sourceProp}"});\n`;
            str += `  ${targetModel}.${hasRel}(${modelName}, { foreignKey: "${sourceProp}"});\n`;
          }
        } else {
            str += `  ${modelName}.belongsTo(${targetModel}, { foreignKey: "${sourceProp}"});\n`;
            str += `  ${targetModel}.${hasRel}(${modelName}, { foreignKey: "${sourceProp}"});\n`;
        }
      }
    });
  });
  return str;
}
```

It is possible that I look silly right now for suggesting it, but I have no knowledge about the code or the abillity to test it at this time. But if this is not a good option because of some edge cases, the answer above could maybe be more usefull.

**Comment by steveschmitt** [comment]
In the case where the "through" table is just for establishing the many-to-many relationship, then it makes sense to remove the other relationships.  Your `PlanHasSpecifications` table may be a good example of this.

But I worry about the case where the "through" table is also a _real entity_, and you might want to query that entity.  For example, suppose `PlanHasSpecifications` had other columns, such as: 

- **date** when the specification was added to the plan, 
- **name** of person who added the specification to the plan,
- **priority** of the specification in the plan, 

I'm just making things up, so it might not make sense in your business, but the idea is that there could be a real `hasMany` and `belongsTo` relationship between `Plan` and `PlanHasSpecification` _in addition to_ the `belongsToMany` relationship between `Plan` and `PlanSpecification`.

Thank you for looking into this.  I will think about it some more.

**Comment by KjellVermeersch** [comment]
I was thinking about this exact issue aswell. 
This made me question how to tackle this as Sequelize intended. According to [this Sequelize doc page](https://sequelize.org/master/manual/advanced-many-to-many.html) you should be able to get and set these columns if there is a belongsToMany association. I myself have never used this specific approach, but I propably will in the future if this is easier. 

It is up to you to choose or find a way to approach this issue I guess. 
As far as my understandig of Sequelize goes, it looks like it is intended to be used with the belongsToMany only. 

At least I tried my best to find out why this is an issue and what possibilities there are to solve it somehow (both for you as for myself).

Keep up the good work! I will propably see the commits comming in when the time is right.

Thanks!

**Comment by steveschmitt** [comment]
Yes, there is more to ponder.  I thought that our original generated code would work like the [Super many-to-many relationship described on that same page](https://sequelize.org/master/manual/advanced-many-to-many.html#the-best-of-both-worlds--the-super-many-to-many-relationship).  But clearly we have a problem that their example does not.

**Comment by KjellVermeersch** [comment]
I totally read over that part, but you are right! It looks like the error doesn't occure when you follow their order. You firstly have to execute the belongsToMany associations, followed by their hasMany' and belongsTo's. The order between the hasMany and belongTo associations respectively don't matter. The belongsToMany just need to be executed first, apparently.

The following code of my example doesn't give an error (I added an extra line to seperate them more clearly):
```JS
// belongsToMany associations need to be executed first!
Plans.belongsToMany(PlanSpecifications, { through: PlanHasSpecifications, foreignKey: "planName", otherKey: "specificationName" });
PlanSpecifications.belongsToMany(Plans, { through: PlanHasSpecifications, foreignKey: "specificationName", otherKey: "planName" });

// Order of below associations doesn't matter
Plans.hasMany(PlanHasSpecifications, { foreignKey: "planName"});
PlanHasSpecifications.belongsTo(Plans, { foreignKey: "planName"});
PlanSpecifications.hasMany(PlanHasSpecifications, { foreignKey: "specificationName"});
PlanHasSpecifications.belongsTo(PlanSpecifications, { foreignKey: "specificationName"});
```

So now the problem is that you will need to make the many-to-many relationships to be added first. 
Maybe you can append every line to an array instead of appending them to a result string. This way you can change the order of every line according to the function. At the and you can join all those strings together with the newline character ('\n').

**Comment by steveschmitt** [comment]
Great!  That makes it much easier.  

Thanks for your work on this.  I'm glad we had this discussion before I tried to fix it with limited knowledge.

**Comment by KjellVermeersch** [comment]
I want to let you know that I created a script that reads the init-models file, reorders the associations and updates the init-models file. So I temporarily resolved this issue by myself and it seems to do the trick for now (still need to fuly test it). 
It really is some ugly code, but it works 😅

**Comment by steveschmitt** [comment]
Cool, glad that you got it working for you.  I'm working now on the code to put the `belongsToMany` first, along with some other minor issues.  It'll be released in the next few days.

**Comment by steveschmitt** [comment]
Fixed in 0.7.4.

Now it puts the belongsToMany relationships first.

**Comment by kevinmamaqi** [comment]
I'm getting the same error even though the belongsToMany is above:
```
location.hasMany(ad, {as: "ads", foreignKey: "locationId"});
media.hasMany(ad, {as: "ads", foreignKey: "mediaId"});
user.hasMany(ad, {as: "ads", foreignKey: "userId"});
```
